### PR TITLE
telemetry: Add telemetry information about builtin functions usage

### LINKED ISSUE #25947: Add builtinfunction usage statistics
## Feature Request

**Is your feature request related to a problem? Please describe:**

Some builtinfunctions doesn't support pushdown. The builtinfunctions’ usage needs to be counted to determine the priority.

**Describe the feature you'd like:**

usage statistics of all the builtinfunctions

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by leiysky**
### What problem does this PR solve?

Issue Number: close #25947

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

Use a `map[string]uint32` to record builtin functions usage information in each telemetry window.

Since for each expression in a single SQL query will be accessed by `expressionRewriter`, we can record the builtin functions usage information as soon as `expressionRewriter.newFunction` is invoked. Then we can submit the result to `GlobalBuiltinFunctionUsageCollector` after executing the query.

After adding this field, the telemetry data will look like:
```json
{
    "beginAt": "2021-07-14T13:42:11.7490453+08:00",
    "executeCount": 0,
    "tiFlashUsage": {
        "pushDown": 0,
        "exchangePushDown": 0
    },
    "coprCacheUsage": {
        "gte0": 0,
        "gte1": 0,
        "gte10": 0,
        "gte20": 0,
        "gte40": 0,
        "gte80": 0,
        "gte100": 0
    },
    "SQLUsage": {
        "total": 0,
        "type": {}
    },
    "builtinFunctionsUsage": {
        "EQInt": 20,
        "EQString": 1,
        "GTInt": 545,
        "GTTime": 548,
        "InString": 1,
        "LTInt": 11,
        "LeftUTF8": 1,
        "PlusInt": 261,
        "SHA2": 1,
        "UnaryNotInt": 1
    }
}
```

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

- Add telemetry information about builtin functions usage

### CODE REVIEW

**Comment by leiysky** [comment]
/cc @kaixu120811

PTAL.

File: telemetry/data_window.go [hunk 1]
```diff
@@ -54,19 +54,21 @@ const (
 	// WindowSize determines how long some data is aggregated by.
 	WindowSize = 1 * time.Hour
 	// SubWindowSize determines how often data is rotated.
-	SubWindowSize = 1 * time.Minute
+	// SubWindowSize = 1 * time.Minute
+	SubWindowSize = 10 * time.Second
```
Comment (tisonkun) [inline telemetry/data_window.go:58]:
```suggestion
	SubWindowSize = 1 * time.Minute
```

**Review by tisonkun — APPROVED** [review]
LGTM.

File: planner/core/expression_rewriter.go [hunk 2]
```diff
@@ -1178,14 +1179,21 @@ func (er *expressionRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok
...
 	if er.disableFoldCounter > 0 {
-		return expression.NewFunctionBase(er.sctx, funcName, retType, args...)
+		ret, err = expression.NewFunctionBase(er.sctx, funcName, retType, args...)
 	}
 	if er.tryFoldCounter > 0 {
-		return expression.NewFunctionTryFold(er.sctx, funcName, retType, args...)
+		ret, err = expression.NewFunctionTryFold(er.sctx, funcName, retType, args...)
 	}
-	return expression.NewFunction(er.sctx, funcName, retType, args...)
+	ret, err = expression.NewFunction(er.sctx, funcName, retType, args...)
```
Comment (LittleFall) [inline planner/core/expression_rewriter.go:1189]: why not `if ... else if ... else ...`? seems without `return`, the original path has been changed.

Comment (leiysky) [reply planner/core/expression_rewriter.go]: Fixed.

**Review by LittleFall — APPROVED** [review]
LGTM

**Comment by LittleFall** [comment]
test need fix:
```
[2021-07-15T08:07:27.464Z] FAIL: planbuilder_test.go:144: testPlanBuilderSuite.TestDisableFold
[2021-07-15T08:07:27.464Z] 
[2021-07-15T08:07:27.464Z] planbuilder_test.go:184:
[2021-07-15T08:07:27.464Z]     c.Assert(rewritenArg, FitsTypeOf, expectedArg)
[2021-07-15T08:07:27.464Z] ... obtained *expression.Constant = &expression.Constant{Value:types.Datum{k:0x4, decimal:0x0, length:0x0, i:-4621415532179685754, collation:"", b:[]uint8(nil), x:interface {}(nil)}, RetType:(*types.FieldType)(0xc00080fc20), DeferredExpr:expression.Expression(nil), ParamMarker:(*expression.ParamMarker)(nil), hashcode:[]uint8(nil), collationInfo:expression.collationInfo{coer:5, coerInit:true, charset:"", collation:"", flen:0}} ("-0.45990349068959124")
[2021-07-15T08:07:27.464Z] ... sample *expression.ScalarFunction = &expression.ScalarFunction{FuncName:model.CIStr{O:"", L:""}, RetType:(*types.FieldType)(nil), Function:expression.builtinFunc(nil), hashcode:[]uint8(nil)} (%!q(PANIC=String method: runtime error: invalid memory address or nil pointer dereference))
```

File: session/session.go [hunk 3]
```diff
@@ -225,6 +225,8 @@ type session struct {
 
 	// indexUsageCollector collects index usage information.
 	idxUsageCollector *handle.SessionIndexUsageCollector
+
+	builtinFunctionUsage telemetry.BuiltinFunctionsUsage
```
Comment (Vancior) [inline session/session.go:229]: This map should be cleared periodically too.

**Review by Vancior — CHANGES_REQUESTED** [review]

File: telemetry/data_window.go [hunk 4]
```diff
@@ -90,6 +91,60 @@ type tiFlashUsageData struct {
...
+		b[scalarFuncSigName] = v + 1
+	}
+}
+
+// Merge BuiltinFunctionsUsage data
+func (b BuiltinFunctionsUsage) Merge(usageData BuiltinFunctionsUsage) {
+	for k, v := range usageData {
+		prev, ok := b[k]
+		if !ok {
+			b[k] = prev
```
[the remark this answers is not in the document]
Comment (leiysky) [reply telemetry/data_window.go]: Fixed.

File: session/session.go (hunk 3 above)
Comment (leiysky) [reply session/session.go:229]: Fixed.

Also, I found we can do the collect work in `session.Close` to reduce lock acquiring.

**Review by Vancior — COMMENTED** [review]
LGTM.
@lucklove

**Review by lucklove — APPROVED** [review]
LGTM

**Review by LittleFall — APPROVED** [review]
LGTM

**Comment by dveeden** [comment]
Please add details about this to https://docs.pingcap.com/tidb/stable/telemetry once merged

**Comment by breeswish** [comment]
@lzmhhh123 is performing additional review for this PR.

**Pull request merged**